roofit/roofitcore/src/RooBinIntegrator.cxx

```diff
@@ -160,69 +156,42 @@ bool RooBinIntegrator::checkLimits() const
+/// Calculate numeric integral at given set of function binding parameters,
+/// for any number of dimensions.
 double RooBinIntegrator::integral(const double *)
 {
   assert(isValid());
+  if (_function->getDimension() < 1) return 0.;
 
   ROOT::Math::KahanSum<double> sum;
 
-  if (_function->getDimension() == 1) {
-    const std::vector<double>& binb = _binb[0];
-
-    for (unsigned int ibin=0; ibin < binb.size() - 1; ++ibin) {
-      const double xhi = binb[ibin + 1];
-      const double xlo = binb[ibin];
-      const double xcenter = (xhi+xlo)/2.;
-      const double binInt = integrand(xvec(xcenter))*(xhi-xlo) ;
-      sum += binInt ;
-    }
-  } else if (_function->getDimension() == 2) {
-    const std::vector<double>& binbx = _binb[0];
-    const std::vector<double>& binby = _binb[1];
-
-    for (unsigned int ibin1=0; ibin1 < binbx.size() - 1; ++ibin1) {
-      const double x1hi = binbx[ibin1 + 1];
-      const double x1lo = binbx[ibin1];
-      double x1center = (x1hi+x1lo)/2 ;
-
-      for (unsigned int ibin2=0; ibin2 < binby.size() - 1; ++ibin2) {
-        const double x2hi = binby[ibin2 + 1];
-        const double x2lo = binby[ibin2];
-        const double x2center = (x2hi+x2lo)/2.;
-
-        const double binInt = integrand(xvec(x1center,x2center))*(x1hi-x1lo)*(x2hi-x2lo) ;
-        sum += binInt ;
-      }
-    }
-  } else if (_function->getDimension() == 3) {
-    const std::vector<double>& binbx = _binb[0];
-    const std::vector<double>& binby = _binb[1];
-    const std::vector<double>& binbz = _binb[2];
-
-    for (unsigned int ibin1=0; ibin1 < binbx.size() - 1; ++ibin1) {
-      const double x1hi = binbx[ibin1 + 1];
-      const double x1lo = binbx[ibin1];
-      double x1center = (x1hi+x1lo)/2 ;
-
-      for (unsigned int ibin2=0; ibin2 < binby.size() - 1; ++ibin2) {
-        const double x2hi = binby[ibin2 + 1];
-        const double x2lo = binby[ibin2];
-        const double x2center = (x2hi+x2lo)/2.;
-
-        for (unsigned int ibin3=0; ibin3 < binbz.size() - 1; ++ibin3) {
-          const double x3hi = binbz[ibin3 + 1];
-          const double x3lo = binbz[ibin3];
-          const double x3center = (x3hi+x3lo)/2.;
-
-          const double binInt = integrand(xvec(x1center,x2center,x3center))*(x1hi-x1lo)*(x2hi-x2lo)*(x3hi-x3lo);
-          sum += binInt ;
-        }
-      }
-    }
-  }
+  recursive_integration(0,1.,{},sum);
 
   return sum.Sum();
 }
 
-
+/**
+ * @brief It performs recursively for loops to perform N-dimensional integration
+ * @param d the current recursivity depth (dimension currently being for-looped)
+ * @param delta the (d-1)-dimensional bin width/area/volume/hypervolume...
+ * @param center the (d-1)-dimensional bin center of the current iteration
+ * @param sum the resulting integral where to accumulate the integral, passed by reference
+ */
+void RooBinIntegrator::recursive_integration(const UInt_t d, const double delta, const std::vector<double> center, ROOT::Math::KahanSum<double>& sum) {
```

Comment:
```suggestion
void RooBinIntegrator::recursive_integration(const UInt_t d, const double delta, std::span<const double> center, ROOT::Math::KahanSum<double>& sum) {
```
Vectors are expensive to copy because of manual memory allocation. Can we use a `std::span` here? Or at least a `const &`?